File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081339Z_claude_claude-opus-5_typescript-nextjs/lib/combat.ts
/**
 * Durable combat sessions: persistence plus the response shapes the combat
 * endpoints return. Ordering rules live in `./initiative`.
 */
import { database } from "./db";
import {
  type InitiativeEntry,
  parseInitiativeEntry,
  sortByInitiative,
} from "./initiative";

export type Condition = { condition: string; remaining_rounds: number };

export type Combatant = InitiativeEntry & { conditions: Condition[] };

export type Session = {
  id: string;
  round: number;
  turn_index: number;
  order: Combatant[];
};

/**
 * Sessions are durable: they live in the `combat_sessions` table. The turn
 * order is stored as a JSON column so a session round-trips as one row.
 */
export function getSession(id: string): Session | undefined {
  const row = database()
    .prepare("SELECT id, round, turn_index, order_json FROM combat_sessions WHERE id = ?")
    .get(id) as Record<string, unknown> | undefined;
  if (!row) return undefined;
  return {
    id: row.id as string,
    round: Number(row.round),
    turn_index: Number(row.turn_index),
    order: JSON.parse(row.order_json as string) as Combatant[],
  };
}

export function hasSession(id: string): boolean {
  return getSession(id) !== undefined;
}

export function saveSession(session: Session): void {
  database()
    .prepare(
      `INSERT INTO combat_sessions (id, round, turn_index, order_json)
       VALUES (?, ?, ?, ?)
       ON CONFLICT(id) DO UPDATE SET
         round = excluded.round,
         turn_index = excluded.turn_index,
         order_json = excluded.order_json`,
    )
    .run(session.id, session.round, session.turn_index, JSON.stringify(session.order));
}

/**
 * Build the combatant list from a request payload, or undefined when invalid.
 * Stricter than a one-shot ordering: a stored session addresses combatants by
 * name (conditions target a name), so names must be non-empty and unique.
 */
export function parseCombatants(value: unknown): Combatant[] | undefined {
  if (!Array.isArray(value) || value.length === 0) return undefined;

  const entries: Combatant[] = [];
  const seen = new Set<string>();
  for (const raw of value) {
    const entry = parseInitiativeEntry(raw);
    if (!entry || entry.name === "") return undefined;
    if (seen.has(entry.name)) return undefined;
    seen.add(entry.name);
    entries.push({ ...entry, conditions: [] });
  }
  return sortByInitiative(entries);
}

export function activeCombatant(session: Session): Combatant {
  return session.order[session.turn_index]!;
}

/** Combatants are exposed by name and score only; dex is an internal tiebreaker. */
export function publicCombatant(combatant: Combatant): { name: string; score: number } {
  return { name: combatant.name, score: combatant.score };
}

export function publicConditions(combatant: Combatant): Condition[] {
  return combatant.conditions.map((entry) => ({ ...entry }));
}

/**
 * Conditions keyed by combatant name. Every combatant is present, with an
 * empty list once their conditions have all expired, so callers can keep
 * reading a stable key per combatant.
 */
export function conditionMap(session: Session): Record<string, Condition[]> {
  const result: Record<string, Condition[]> = {};
  for (const combatant of session.order) {
    result[combatant.name] = publicConditions(combatant);
  }
  return result;
}

/** The canonical session body, shared by session create / advance responses. */
export function sessionState(session: Session): Record<string, unknown> {
  return {
    id: session.id,
    round: session.round,
    turn_index: session.turn_index,
    active: publicCombatant(activeCombatant(session)),
    order: session.order.map(publicCombatant),
    conditions: conditionMap(session),
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081336Z_claude_claude-opus-5_go-stdlib/combat.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"
	"sync"
)

// Stateful combat tracker: an initiative order per session, a turn cursor, and
// per-combatant conditions that tick down. Sessions live in memory and are
// mirrored to SQLite after every mutation (see store.go).

type conditionState struct {
	Condition       string `json:"condition"`
	RemainingRounds int    `json:"remaining_rounds"`
}

// combatSession is one tracked fight. TurnIndex always points at a valid entry
// of Order, which is never empty — handlers rely on that to report the active
// combatant without a bounds check, and the loader drops any persisted row that
// would violate it.
type combatSession struct {
	ID         string
	Round      int
	TurnIndex  int
	Order      []initiativeEntry
	Conditions map[string][]conditionState
}

type combatStore struct {
	mu       sync.Mutex
	sessions map[string]*combatSession
}

var sessions = &combatStore{sessions: map[string]*combatSession{}}

// conditionsView deep-copies the condition lists so the response cannot alias
// session state once the lock is released. Combatants whose conditions have all
// expired keep an empty list, which is how callers tell "cleared" apart from
// "never present".
func (s *combatSession) conditionsView() map[string][]conditionState {
	view := make(map[string][]conditionState, len(s.Conditions))
	for name, list := range s.Conditions {
		view[name] = append(make([]conditionState, 0, len(list)), list...)
	}
	return view
}

// hasCombatant reports whether name appears in the initiative order.
func (s *combatSession) hasCombatant(name string) bool {
	for _, e := range s.Order {
		if e.Name == name {
			return true
		}
	}
	return false
}

// ---------- payloads ----------

type createSessionRequest struct {
	ID         *string     `json:"id"`
	Combatants []combatant `json:"combatants"`
}

// sessionResponse is the create reply: no conditions, since a new session has
// none.
type sessionResponse struct {
	ID        string            `json:"id"`
	Round     int               `json:"round"`
	TurnIndex int               `json:"turn_index"`
	Active    initiativeEntry   `json:"active"`
	Order     []initiativeEntry `json:"order"`
}

// sessionStateResponse is the full session view returned by reads and advances.
type sessionStateResponse struct {
	ID         string                      `json:"id"`
	Round      int                         `json:"round"`
	TurnIndex  int                         `json:"turn_index"`
	Active     initiativeEntry             `json:"active"`
	Order      []initiativeEntry           `json:"order"`
	Conditions map[string][]conditionState `json:"conditions"`
}

type addConditionRequest struct {
	Target         *string          `json:"target"`
	Condition      *string          `json:"condition"`
	DurationRounds *json.RawMessage `json:"duration_rounds"`
}

type addConditionResponse struct {
	Target     string           `json:"target"`
	Conditions []conditionState `json:"conditions"`
}

// stateOf renders the full session view. Callers must hold sessions.mu.
func (s *combatSession) stateOf() sessionStateResponse {
	return sessionStateResponse{
		ID:         s.ID,
		Round:      s.Round,
		TurnIndex:  s.TurnIndex,
		Active:     s.Order[s.TurnIndex],
		Order:      s.Order,
		Conditions: s.conditionsView(),
	}
}

// ---------- POST /v1/combat/sessions ----------

func handleCreateSession(w http.ResponseWriter, r *http.Request) {
	var req createSessionRequest
	if err := decodeBody(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	if req.ID == nil || strings.TrimSpace(*req.ID) == "" {
		writeError(w, http.StatusBadRequest, "id is required")
		return
	}
	if len(req.Combatants) == 0 {
		writeError(w, http.StatusBadRequest, "at least one combatant is required")
		return
	}
	order, msg := initiativeOrder(req.Combatants, true)
	if msg != "" {
		writeError(w, http.StatusBadRequest, msg)
		return
	}

	// The stored id is the raw value, not the trimmed one used for the blank
	// check, so lookups keep matching what the client sent.
	id := *req.ID
	session := &combatSession{
		ID:         id,
		Round:      1,
		TurnIndex:  0,
		Order:      order,
		Conditions: map[string][]conditionState{},
	}

	// Creating a session is idempotent: re-posting an id rolls that encounter
	// back to round 1 with a fresh initiative order. Combat state is persisted,
	// so rejecting a repeat id would make a restarted server refuse to start the
	// same encounter again.
	sessions.mu.Lock()
	sessions.sessions[id] = session
	sessions.mu.Unlock()
	flush()

	writeJSON(w, http.StatusOK, sessionResponse{
		ID:        session.ID,
		Round:     session.Round,
		TurnIndex: session.TurnIndex,
		Active:    session.Order[session.TurnIndex],
		Order:     session.Order,
	})
}

// ---------- GET /v1/combat/sessions/{id} ----------

func handleGetSession(w http.ResponseWriter, r *http.Request) {
	sessions.mu.Lock()
	defer sessions.mu.Unlock()
	session, ok := sessions.sessions[r.PathValue("id")]
	if !ok {
		writeError(w, http.StatusNotFound, "session not found")
		return
	}
	writeJSON(w, http.StatusOK, session.stateOf())
}

// ---------- POST /v1/combat/sessions/{id}/conditions ----------

func handleAddCondition(w http.ResponseWriter, r *http.Request) {
	var req addConditionRequest
	if err := decodeBody(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	if req.Target == nil || *req.Target == "" {
		writeError(w, http.StatusBadRequest, "target is required")
		return
	}
	if req.Condition == nil || *req.Condition == "" {
		writeError(w, http.StatusBadRequest, "condition is required")
		return
	}
	duration, ok := asInt(req.DurationRounds)
	if !ok {
		writeError(w, http.StatusBadRequest, "duration_rounds must be an integer")
		return
	}
	if duration < 1 {
		writeError(w, http.StatusBadRequest, "duration_rounds must be positive")
		return
	}

	sessions.mu.Lock()
	session, found := sessions.sessions[r.PathValue("id")]
	if !found {
		sessions.mu.Unlock()
		writeError(w, http.StatusNotFound, "session not found")
		return
	}
	if !session.hasCombatant(*req.Target) {
		sessions.mu.Unlock()
		writeError(w, http.StatusBadRequest, "target is not a combatant in this session")
		return
	}

	// Re-applying an existing condition refreshes its duration rather than
	// stacking a second copy.
	list := session.Conditions[*req.Target]
	refreshed := false
	for i := range list {
		if list[i].Condition == *req.Condition {
			list[i].RemainingRounds = duration
			refreshed = true
			break
		}
	}
	if !refreshed {
		list = append(list, conditionState{Condition: *req.Condition, RemainingRounds: duration})
	}
	session.Conditions[*req.Target] = list
	view := append([]conditionState(nil), list...)
	sessions.mu.Unlock()

	// flush takes every store's lock, so it must run unlocked.
	flush()

	writeJSON(w, http.StatusOK, addConditionResponse{Target: *req.Target, Conditions: view})
}

// ---------- POST /v1/combat/sessions/{id}/advance ----------

func handleAdvanceTurn(w http.ResponseWriter, r *http.Request) {
	sessions.mu.Lock()
	session, ok := sessions.sessions[r.PathValue("id")]
	if !ok {
		sessions.mu.Unlock()
		writeError(w, http.StatusNotFound, "session not found")
		return
	}

	session.TurnIndex++
	if session.TurnIndex >= len(session.Order) {
		session.TurnIndex = 0
		session.Round++
	}
	session.tickConditions(session.Order[session.TurnIndex].Name)

	resp := session.stateOf()
	sessions.mu.Unlock()
	flush()

	writeJSON(w, http.StatusOK, resp)
}

// tickConditions decrements the named combatant's conditions and drops the
// expired ones. Durations are consumed at the start of that combatant's own
// turn, so a 1-round condition applied on their turn survives until it comes
// around again. The (possibly empty) map entry is kept so callers can see that
// conditions were cleared rather than never present.
func (s *combatSession) tickConditions(name string) {
	list := s.Conditions[name]
	if len(list) == 0 {
		return
	}
	kept := list[:0]
	for _, c := range list {
		c.RemainingRounds--
		if c.RemainingRounds > 0 {
			kept = append(kept, c)
		}
	}
	s.Conditions[name] = kept
}
